data/libs/EquipSet.lua

```diff
@@ -1,404 +1,643 @@
+--
+--  slot  - optional HullConfig.Slot, the slot the equipment item is being
+--          validated against. If not present, the function validates that the
+--          passed equipment item is a non-slot equipment item.
+--
+---@param equip EquipType
+---@param slot HullConfig.Slot?
+function EquipSet.CompatibleWithSlot(equip, slot)
+	local equipSlot = equip.slot or false
+	if not slot then return not equipSlot end
+
+	return equipSlot and
+		slotTypeMatches(equip.slot.type, slot.type)
+		and (equip.slot.size <= slot.size)
+		and (equip.slot.size >= (slot.size_min or slot.size))
+end
+
+-- Constructor: New
+--
+-- Construct a new EquipSet object for the given ship.
+---@param ship Ship
+function EquipSet:Constructor(ship)
+	self.ship = ship
+	self.config = HullConfig.GetHullConfigs()[ship.shipId]
+
+	-- Stores a mapping of slot id -> equipment item
+	-- Non-slot equipment is stored in the array portion.
+	self.installed = {} ---@type table<string|integer, EquipType>
+	-- NOTE: the integer value stored in the cache is NOT the current array
+	-- index of the given item. It's simply a non-nil integer to indicate the
+	-- item is not installed in a slot.
+	self.cache = {} ---@type table<EquipType, string|integer>
+	-- This provides a unique index value for non-slot equipment cache entries.
+	-- It is monotonically increasing and prevents ID collisions for recursive
+	-- slots provided by non-slot equipment items.
+	self.cacheIndex = 1
+
+	-- Stores a mapping of slot id -> slot handle
+	-- Simplifies slot lookup for slots defined on equipment items
+	self.slotCache = {} ---@type table<string, HullConfig.Slot>
+	-- Stores the inverse mapping for looking up the compound id of a slot by
+	-- the slot object itself.
+	self.idCache = {} ---@type table<HullConfig.Slot, string>
+
+	self:BuildSlotCache()
+
+	-- List of listener functions to be notified of changes to the set of
+	-- installed equipment in this EquipSet
+	self.listeners = {} ---@type EquipSet.Listener[]
+
+	-- Initialize ship properties we're responsible for modifying
+	self.ship:setprop("mass_cap", self.ship["mass_cap"] or 0)
+	self.ship:setprop("equipVolume", self.ship.equipVolume or 0)
+	self.ship:setprop("totalVolume", self.ship.totalVolume or self.config.capacity)
 end
 
-function EquipSet:CallListener(slot)
-	if not listeners[self] then return end
+-- Callback: OnShipTypeChanged
+--
+-- Called when the type of the owning ship is changed. Removes all installed
+-- equipment and tries to leave the ship in an "empty" state.
+function EquipSet:OnShipTypeChanged()
+	---@type (string|integer)[]
+	local to_remove = {}
 
-	for _, listener in ipairs(listeners[self]) do
-		listener(slot)
+	for k in pairs(self.installed) do
+		table.insert(to_remove, k)
 	end
-end
 
-function EquipSet:Serialize()
-	local serialize = {
-		slots = {}
-	}
+	-- Sort the longest strings first, as equipment installed in subslots will need to be removed first
+	table.sort(to_remove, function(a, b)
```

Comment:
Thinking-out-loud: (no actual change required)

Wouldn't it be neater to have each equipment handle its own removal logic in the `OnRemove` callback? If an equipment item has sub-slots, it should be responsible for clearing those out instead of having to have fairly complex logic here (and potentially elsewhere) to ensure proper cleanup. 

To avoid code duplication, you can make a hierarchy of equipment classes - BasicEquipment, SlotEquipment, etc.  EquipSet itself could just be a specialisation of SlotEquipment then.

Currently, `this.Remove()` does not recurse into sub-slots and requires the caller to have already removed all sub-items, which I guess this function does. Yes I know its documented, but future modders/maintainers may not read that and leave dangling sub-items.

Reply:
(politely) This is an OOP trap. I don't want to have a SlotEquipment and a PassengerEquipment and then worry about a PassengerSlotEquipment... you get where I'm going with this.

On topic, `Remove()` does not recurse for the very simple reason that recursively removing an item "vanishes" the items installed in sub slots. Callers would need to do their own pre-traversal of the slots provided by an item if they want to retain the items that are installed, at which point you might as well post-order call Remove during that traversal.

It is definitely something I could consider supporting in the future if the need arises, but so far I haven't seen a significant pain-point that a few lines of code hasn't solved.

Reply:
I just like APIs which are close to idiot-proof (M <- an idiot..)

That said, I'm not big into OO, my go-to language is C.. but sometimes having classes is nice.

Anyway, was just a thought. Thanks for your informative reply; closing.

Reply:
I should have spoken a bit more carefully - the humor module was malfunctioning this morning.

Other than the deadly-diamond OO pattern I was alluding to, there is definitely value in having the equipment handle some actions of its own, and it's something I expect we'll need to do quite soon when we get into "cabin equipment" which provides (shielded) cargo space or additional crew quarters / equipment volume.

For the most part, that behavior should all be able to go into the base `EquipType` (which does follow a "class" pattern) - this allows each individual equipment "class" to subscribe to the needed parts of that behavior in a compositional manner, by simply setting a data value in the equipment definition.